streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -441,46 +442,46 @@ private void closeOpenIterators() {
         }
     }
 
-    private class RocksDbIterator implements KeyValueIterator<Bytes, byte[]> {
+    private class RocksDbIterator extends AbstractIterator<KeyValue<Bytes, byte[]>> implements KeyValueIterator<Bytes, byte[]> {
         private final String storeName;
         private final RocksIterator iter;
 
         private volatile boolean open = true;
 
+        private KeyValue<Bytes, byte[]> next;
+
         RocksDbIterator(final String storeName,
                         final RocksIterator iter) {
             this.iter = iter;
             this.storeName = storeName;
         }
 
-        byte[] peekRawKey() {
-            return iter.key();
-        }
-
-        private KeyValue<Bytes, byte[]> getKeyValue() {
-            return new KeyValue<>(new Bytes(iter.key()), iter.value());
-        }
-
         @Override
         public synchronized boolean hasNext() {
             if (!open) {
                 throw new InvalidStateStoreException(String.format("RocksDB store %s has closed", storeName));
             }
```

Comment:
Having this check here has got me to thinking more about this issue.  

Without this guard condition, we have some failing unit tests. 

In both the `RocksDBIterator` and the `AbstractIterator` all calls to `next()` make a call to `hasNext()` first before returning the next object.  I'm not sure about changing the semantics where we return from `next()` without calling `hasNext()` first (which if we end up keeping those semantics, leaves us in the same position as before extending `AbstractIterator`).  

I could be overthinking this but I'm not entirely comfortable with returning a value from `next()` after the store has been explicitly closed.  I feel like that creates more corner cases for potential errors or unexpected behavior.

WDYT?

Reply:
@bbejeck That is a good question!

Originally I thought it is okay to always calling `hasNext` inside `next()`, as long as we make sure `hasNext` implementation is idempotent, i.e. calling it multiple times before `next()` does not have side effect is sufficient. But by making it idempotent we could have the corner case you mentioned. For example:

```
t0: call `hasNext()` -> store is still open -> call `makeNext` -> `next` field is set.
t1: store is closed.
t2: call `next()` -> call `hasNext()` again
```

Without this check, at `t3` we would still return the `next` field.